app/api/contests/get_qualification_state.go

```diff
@@ -0,0 +1,236 @@
+		ContestMaxTeamSize       int32
+		ContestEnteringCondition *string
+	}
+	err = srv.Store.Items().VisibleByID(user, itemID).Where("items.duration IS NOT NULL").
+		Select("items.has_attempts, items.contest_max_team_size, items.contest_entering_condition").
+		Take(&contestInfo).Error()
+	if gorm.IsRecordNotFoundError(err) {
+		return service.InsufficientAccessRightsError
+	}
+	service.MustNotBeError(err)
+
+	if apiError := srv.checkGroupIDForGetQualificationState(groupID, itemID, contestInfo.IsTeamOnly, user); apiError != service.NoError {
+		return apiError
+	}
+
+	alreadyStarted, err := srv.Store.ContestParticipations().
+		Where("contest_item_id = ?", itemID).
+		Where("group_id = ?", groupID).
+		Where("contest_started_at IS NOT NULL").HasRows()
+	service.MustNotBeError(err)
+
+	membersCount, members, currentUserCanEnter, qualifiedMembersCount :=
+		srv.getQualificatonInfo(contestInfo.IsTeamOnly, groupID, itemID, user)
+
+	qualificationState := computeQualificationState(
+		alreadyStarted, contestInfo.IsTeamOnly, contestInfo.ContestMaxTeamSize,
+		contestInfo.ContestEnteringCondition, membersCount, qualifiedMembersCount)
+
+	result := &contestGetQualificationStateResponse{
+		State:               qualificationState,
+		EnteringCondition:   contestInfo.ContestEnteringCondition,
+		CurrentUserCanEnter: currentUserCanEnter,
+		OtherMembers:        members,
+	}
+	if contestInfo.IsTeamOnly {
+		result.MaxTeamSize = &contestInfo.ContestMaxTeamSize
+	}
+	render.Respond(w, r, result)
+	return service.NoError
+}
+
+func (srv *Service) checkGroupIDForGetQualificationState(groupID, itemID int64, isTeamOnly bool, user *database.User) service.APIError {
+	if isTeamOnly {
+		var teamGroupID int64
+		err := srv.Store.Groups().TeamGroupForTeamItemAndUser(itemID, user).
+			PluckFirst("groups.id", &teamGroupID).Error()
+		if gorm.IsRecordNotFoundError(err) {
+			return service.InsufficientAccessRightsError
+		}
+		service.MustNotBeError(err)
+		if teamGroupID != groupID {
+			return service.InsufficientAccessRightsError
+		}
+	} else if user.SelfGroupID == nil || groupID != *user.SelfGroupID {
+		return service.InsufficientAccessRightsError
+	}
+	return service.NoError
+}
+
+func computeQualificationState(alreadyStarted, isTeamOnly bool, maxTeamSize int32, contestEnteringCondition *string,
+	membersCount, qualifiedMembersCount int32) string {
+	var qualificationState string
+	if alreadyStarted {
+		qualificationState = "already_started"
+	} else {
+		qualificationState = "ready"
+		if isTeamOnly && maxTeamSize < membersCount ||
+			contestEnteringCondition != nil &&
+				!isContestEnteringConditionSatisfied(*contestEnteringCondition, membersCount, qualifiedMembersCount) {
+			qualificationState = "not_ready"
+		}
+	}
+	return qualificationState
+}
+
+func isContestEnteringConditionSatisfied(contestEnteringCondition string, membersCount, qualifiedMembersCount int32) bool {
+	return !(contestEnteringCondition == "All" &&
```

Comment:
Multiple negations make things complicated to verify.
I would prefer something trivial to understand, check and modify such as :
```
return contestEnteringCondition == "None" || 
    (contestEnteringCondition == "One" && qualifiedMembersCount >= 1) ||
    (contestEnteringCondition == "Half" && qualifiedMembersCount >= math.Ceil(membersCount/2.0))
    (contestEnteringCondition == "All" && qualifiedMembersCount == membersCount)
```

Reply:
fixed